File: LeetCodeSolutions/LeetCode_0095.py
# Definition for a binary tree node.
from typing import List


class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def generateTrees(self, n: int) -> List[TreeNode]:
        """
        卡特兰数
        Catalan: https://oi-wiki.org/math/catalan/
        """
        if not n:
            return []

        def helper(start, end):
            if start > end:
                return [None, ]
            all_tree = []
            for i in range(start, end + 1):
                left = helper(start, i - 1)
                right = helper(i + 1, end)

                for l in left:
                    for r in right:
                        root = TreeNode(i)
                        root.left = l
                        root.right = r
                        all_tree.append(root)
            return all_tree

        return helper(1, n)

File: LeetCodeSolutions/LeetCode_0095.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<TreeNode*> generateTrees(int n) {
        if (n == 0) return {};
        return helper(0, n + 1);
    }

    vector<TreeNode*> helper(int lo, int hi){
        if (lo + 1 >= hi) return {nullptr};
        vector<TreeNode*> ret;
        for(int val = lo + 1; val < hi; ++val){
            auto left = helper(lo, val);
            auto right = helper(val, hi);
            for(auto l: left){
                for(auto r: right){
                    auto root = new TreeNode(val);
                    root->left = l;
                    root->right = r;
                    ret.push_back(root);
                }
            }
        }
        return ret;
    }
};